// react-cosmos-ui/dist/testHelpers/pluginMocks.js
import { getPluginContext, } from 'react-plugin';
import { getMethodsOf, mockMethodsOf, on } from './pluginHelpers.js';
export function getRouterContext() {
    return getPluginContext('router');
}
export function getMessageHandlerContext() {
    return getPluginContext('messageHandler');
}
export function getRendererCoreContext() {
    return getPluginContext('rendererCore');
}
export function getRouterMethods() {
    return getMethodsOf('router');
}
export function getMessageHandlerMethods() {
    return getMethodsOf('messageHandler');
}
export function getCoreMethods() {
    return getMethodsOf('core');
}
export function getRendererCoreMethods() {
    return getMethodsOf('rendererCore');
}
export function getNotificationsMethods() {
    return getMethodsOf('notifications');
}
export function getRendererPreviewMethods() {
    return getMethodsOf('rendererPreview');
}
export function mockStorage(methods = {}) {
    const allMethods = {
        loadCache: jest.fn(),
        getItem: jest.fn(),
        setItem: jest.fn(),
        ...methods,
    };
    mockMethodsOf('storage', allMethods);
    return allMethods;
}
export function mockRouter(methods = {}) {
    const allMethods = {
        getSelectedFixtureId: jest.fn(),
        selectFixture: jest.fn(),
        unselectFixture: jest.fn(),
        ...methods,
    };
    mockMethodsOf('router', allMethods);
    return allMethods;
}
export function mockCore(methods = {}) {
    const allMethods = {
        registerCommands: () => jest.fn(),
        runCommand: () => jest.fn(),
        getProjectId: jest.fn(),
        getFixtureFileVars: jest.fn(),
        isDevServerOn: jest.fn(),
        getWebRendererUrl: jest.fn(),
        ...methods,
    };
    mockMethodsOf('core', allMethods);
    return allMethods;
}
export function mockMessageHandler(methods = {}) {
    const allMethods = {
        postRendererRequest: jest.fn(),
        ...methods,
    };
    mockMethodsOf('messageHandler', allMethods);
    return allMethods;
}
export function mockRendererCore(methods = {}) {
    const allMethods = {
        getConnectedRendererIds: jest.fn(),
        getPrimaryRendererId: jest.fn(),
        getFixtures: jest.fn(),
        getFixtureState: jest.fn(),
        isRendererConnected: jest.fn(),
        setFixtureState: jest.fn(),
        selectPrimaryRenderer: jest.fn(),
        receiveResponse: jest.fn(),
        ...methods,
    };
    mockMethodsOf('rendererCore', allMethods);
    return allMethods;
}
export function mockRendererPreview(methods = {}) {
    const allMethods = {
        getUrlStatus: jest.fn(),
        getRuntimeStatus: jest.fn(),
        ...methods,
    };
    mockMethodsOf('rendererPreview', allMethods);
    return allMethods;
}
export function mockNotifications(methods = {}) {
    const allMethods = {
        pushStickyNotification: jest.fn(),
        removeStickyNotification: jest.fn(),
        pushTimedNotification: jest.fn(),
        ...methods,
    };
    mockMethodsOf('notifications', allMethods);
    return allMethods;
}
export function mockFixtureTree(methods = {}) {
    const allMethods = {
        revealFixture: jest.fn(),
        ...methods,
    };
    mockMethodsOf('fixtureTree', allMethods);
    return allMethods;
}
export function onRouter(events = {}) {
    const allEvents = {
        fixtureChange: jest.fn(),
        ...events,
    };
    on('router', allEvents);
    return allEvents;
}
export function onMessageHandler(events = {}) {
    const allEvents = {
        serverMessage: jest.fn(),
        rendererResponse: jest.fn(),
        ...events,
    };
    on('messageHandler', allEvents);
    return allEvents;
}
export function onRendererCore(events = {}) {
    const allEvents = {
        request: jest.fn(),
        response: jest.fn(),
        ...events,
    };
    on('rendererCore', allEvents);
    return allEvents;
}
