// react-cosmos-ui/dist/plugins/ResponsivePreview/index.js
import React from 'react';
import { createPlugin } from 'react-plugin';
import { ResponsivePreview } from './ResponsivePreview/ResponsivePreview.js';
import { DEFAULT_DEVICES, DEFAULT_VIEWPORT_STATE, VIEWPORT_STORAGE_KEY, } from './shared.js';
import { ToggleButton } from './ToggleButton/index.js';
const { plug, namedPlug, register } = createPlugin({
    name: 'responsivePreview',
    defaultConfig: {
        devices: DEFAULT_DEVICES,
    },
});
plug('rendererPreviewOuter', ({ children, pluginContext }) => {
    const { getConfig } = pluginContext;
    const { devices } = getConfig();
    const { enabled, viewport, scaled } = getViewportState(pluginContext);
    const onViewportChange = useViewportChange(pluginContext);
    const onScaledChange = useScaledChange(pluginContext);
    return (React.createElement(ResponsivePreview, { devices: devices, enabled: enabled, viewport: viewport, scaled: scaled, setViewport: onViewportChange, setScaled: onScaledChange }, children));
});
namedPlug('rendererAction', 'responsivePreview', ({ pluginContext }) => {
    const { getMethodsOf } = pluginContext;
    const core = getMethodsOf('core');
    const viewportState = getViewportState(pluginContext);
    const { enabled, viewport } = viewportState;
    if (!core.getWebRendererUrl())
        return null;
    return (React.createElement(ToggleButton, { selected: enabled, onToggle: () => {
            if (enabled) {
                setViewportState(pluginContext, { ...viewportState, enabled: false });
                setFixtureStateViewport(pluginContext, null);
            }
            else {
                setViewportState(pluginContext, { ...viewportState, enabled: true });
                setFixtureStateViewport(pluginContext, viewport);
            }
        } }));
});
export { register };
if (process.env.NODE_ENV !== 'test')
    register();
function useViewportChange(context) {
    const viewportState = getViewportState(context);
    return React.useCallback((viewportChange) => {
        const viewport = typeof viewportChange === 'function'
            ? viewportChange(viewportState.viewport)
            : viewportChange;
        setViewportState(context, { ...viewportState, enabled: true, viewport });
        setFixtureStateViewport(context, viewport);
    }, [context, viewportState]);
}
function useScaledChange(context) {
    const viewportState = getViewportState(context);
    return React.useCallback((scaled) => setViewportState(context, { ...viewportState, scaled }), [context, viewportState]);
}
function getViewportState(context) {
    const { getMethodsOf } = context;
    const storage = getMethodsOf('storage');
    const viewportState = storage.getItem(VIEWPORT_STORAGE_KEY) ||
        DEFAULT_VIEWPORT_STATE;
    const rendererCore = getMethodsOf('rendererCore');
    const fixtureState = rendererCore.getFixtureState();
    return fixtureState.viewport
        ? { ...viewportState, enabled: true, viewport: fixtureState.viewport }
        : viewportState;
}
function setViewportState(context, viewportState) {
    const { getMethodsOf } = context;
    const storage = getMethodsOf('storage');
    storage.setItem(VIEWPORT_STORAGE_KEY, viewportState);
}
function setFixtureStateViewport(context, viewport) {
    const { getMethodsOf } = context;
    const rendererCore = getMethodsOf('rendererCore');
    rendererCore.setFixtureState(fixtureState => ({
        ...fixtureState,
        viewport,
    }));
}
