// react-cosmos-core/dist/playground/playgroundUrl.js
export function stringifyPlaygroundUrlQuery(urlParams) {
    const encodedUrlParams = {};
    if (urlParams.fixtureId) {
        encodedUrlParams.fixtureId = JSON.stringify(urlParams.fixtureId);
    }
    return stringifyUrlQuery(encodedUrlParams);
}
export function parsePlaygroundUrlQuery(query) {
    const encodedUrlParams = parseUrlQuery(query);
    const decoded = {};
    if (encodedUrlParams.fixtureId) {
        decoded.fixtureId = JSON.parse(encodedUrlParams.fixtureId);
    }
    return decoded;
}
export function stringifyRendererUrlQuery(urlParams) {
    const encodedUrlParams = {};
    if (urlParams._fixtureId) {
        encodedUrlParams._fixtureId = JSON.stringify(urlParams._fixtureId);
    }
    return stringifyUrlQuery(encodedUrlParams);
}
export function parseRendererUrlQuery(query) {
    const encodedUrlParams = parseUrlQuery(query);
    const decoded = {};
    if (encodedUrlParams._fixtureId) {
        decoded._fixtureId = JSON.parse(encodedUrlParams._fixtureId);
    }
    return decoded;
}
export function parseUrlQuery(query) {
    return Object.fromEntries(new URLSearchParams(query));
}
export function stringifyUrlQuery(params) {
    return new URLSearchParams(params).toString();
}
